format/applebookmark/apple_bookmark.go

```diff
@@ -0,0 +1,454 @@
+
+	d.FieldBool("supports_read_dir_attr")
+	d.FieldBool("supports_copy_file")
+	d.FieldBool("supports_deny_modes")
+	d.FieldBool("supports_symbolic_links")
+	d.FieldBool("reserved_2")
+	d.FieldBool("supports_exchange")
+	d.FieldBool("supports_search_fs")
+	d.FieldBool("supports_persistent_ids")
+
+	d.FieldBool("supports_extended_security")
+	d.FieldBool("has_no_root_directory_times")
+	d.FieldBool("supports_flock")
+	d.FieldBool("supports_case_preserved_names")
+	d.FieldBool("supports_case_sensitive_names")
+	d.FieldBool("supports_fast_stat_fs")
+	d.FieldBool("supports_rename")
+	d.FieldBool("supports_journaling")
+
+	d.FieldBool("supports_zero_runs")
+	d.FieldBool("supports_sparse_files")
+	d.FieldBool("is_journaling")
+	d.FieldBool("reserved_3")
+	d.FieldBool("supports_path_from_id")
+	d.FieldBool("supports_mandatory_byte_range_locks")
+	d.FieldBool("supports_hard_links")
+	d.FieldBool("supports_2_tb_file_size")
+
+	d.FieldRawLen("reserved_4", 3)
+	d.FieldBool("has64_bit_object_ids")
+	d.FieldBool("supports_decmp_fs_compression")
+	d.FieldBool("supports_hidden_files")
+	d.FieldBool("supports_remote_events")
+	d.FieldBool("supports_volume_sizes")
+}
+
+func decodePropertyFlags(d *decode.D, bitFn func(d *decode.D)) {
+	d.FieldStruct("property_flags", bitFn)
+
+	d.FieldStruct("enabled_property_flags", bitFn)
+}
+
+var cocoaTimeEpochDate = time.Date(2001, time.January, 1, 0, 0, 0, 0, time.UTC)
+
+type tocHeader struct {
+	nextTOCOffset    uint64
+	numEntries       uint64
+	entryArrayOffset int64
+}
+
+func (hdr *tocHeader) decodeEntries(d *decode.D) {
+	for k := uint64(0); k < hdr.numEntries; k++ {
+		entry := new(tocEntry)
+
+		d.FieldStruct("entry", func(d *decode.D) {
+			entry.key = d.FieldU32("key", elementTypeMap)
+
+			// if the key has the top bit set, then (key & 0x7fffffff)
+			// gives the offset of a string record.
+			if entry.key&0x80000000 != 0 {
+				d.FieldStruct("key_string", func(d *decode.D) {
+					d.SeekAbs(calcOffset(entry.key&0x7fffffff), decodeRecord)
+				})
+			}
+
+			entry.recordOffset = calcOffset(d.FieldU32("offset_to_record"))
+
+			d.FieldU32("unused")
+
+			switch entry.key {
+			case elementTypeTargetFlags:
+				d.SeekAbs(entry.recordOffset, resourcePropDecoder.decode)
+			case elementTypeVolumeFlags:
+				d.SeekAbs(entry.recordOffset, volumePropDecoder.decode)
+			default:
+				d.SeekAbs(entry.recordOffset, decodeRecord)
+			}
+
+		})
+	}
+}
+
+func decodeTOCHeader(d *decode.D) *tocHeader {
+	hdr := new(tocHeader)
+
+	d.FieldStruct("toc_header", func(d *decode.D) {
+		d.FieldU32("toc_size")
+		d.FieldU32("magic", d.AssertU(0xfffffffe))
+		d.FieldU32("identifier")
+		hdr.nextTOCOffset = d.FieldU32("next_toc_offset")
+		hdr.numEntries = d.FieldU32("num_entries_in_toc")
+		hdr.entryArrayOffset = d.Pos()
+	})
+
+	return hdr
+}
+
+type tocEntry struct {
+	key          uint64
+	recordOffset int64
+}
+
+const (
+	arrayEntrySize = 4
+	dictEntrySize  = 4
+)
+
+type stack []int64
+
+var offsetStack stack
```

Comment:
Yeah, I don't know what I was thinking making that global. Also a problem in `pushAndPop`, the method references the global variable instead of the struct pointer.

Reply:
> Yeah, I don't know what I was thinking making that global. Also a problem in pushAndPop, the method references the global variable instead of the struct pointer.

No worries. Im surprised it was not detect by the test running with -race hmm

>After some thought the cleanest way to do things was to make the stack internal to the decode.D. Maybe this is good going forward? If so, can fix up the bplist decoder to use it as well since it does something similar.

I'm a bit reluctant to add things to `*decode.D`, need to think a bit more about it if so. I guess that field would have to get inherited down into sub decoders? could we go with a separate type for this for now? maybe we could put it in the decode package if it should be shared? another option is to put bplist and apple_bookmark into formart/apple etc so they can share things?

What do you think about something like this:

```go
```

Reply:
For some reason it is not catching the infinite recursion in the example that I produced with your `fq` one-liner there. See the `loop.book` that I added to the testdata.

Reply:
> For some reason it is not catching the infinite recursion in the example that I produced with your fq one-liner there. See the loop.book that I added to the testdata.

Was a typo, see code comment

Reply:
> This is really cool, I was trying to figure out the cleanest way to add the stack to the scope of the decodeRecord function that didn't make a huge mess when passing parameters around, but I didn't know closures could be used like this!

Yeap very useful pattern! go has quite good support for closures and lambdas, being GC:d and blurring what is stack and heap helps also. The syntax is a bit verbose, but not sure i wish for ml/haskell syntax either, a tiny bit more syntax help would be nice :)